SSA/Projects/CIRCT/Handshake/Handshake.lean

```diff
@@ -112,11 +115,71 @@ def altMerge (x y : Stream α) : Stream α :=
           | none   => .right
         (y0, x, y, nextConsume)
 
-end Stream
+-- this is basically the same as the fork in DC
+def fork (x : Stream α) : Stream α × Stream α :=
+  Stream.corec₂ (β := Stream α) x
+    fun x => Id.run <| do
+      let x0 := x 0
+      let x' := x.tail
+      (x0, x0, x')
+
+-- not entirely nondeterministic (still picks left first)
+def controlMerge (x y : Stream α) : Stream α × Stream Bool :=
+  Stream.corec₂ (β := Stream α × Stream α) (x, y) fun ⟨x, y⟩ =>
+    match x 0, y 0 with
+    | some x', some _ => (some x', some true, (x.tail, y))
+    | some x', none => (some x', some true, (x.tail, y.tail))
+    | none, some y' => (some y', some false, (x.tail, y.tail))
+    | none, none => (none, none, (x.tail, y.tail))
+
+
+def join (x y : Stream α) : Stream Unit :=
+  Stream.corec (β := Stream α × Stream α) (x, y) fun ⟨x, y⟩ =>
+    match x 0, y 0 with
+    | some _, some _ => (some (), (x.tail, y.tail))
+    | some _, none   => (none, (x, y.tail))
+    | none, some _   => (none, (x.tail, y))
+    | none, none     => (none, (x.tail, y.tail))
+
+-- select stream and two inputs (deterministic mergs)
+def mux (x y : Stream α) (c : Stream Bool) : Stream α :=
+  Stream.corec (β := Stream α × Stream α × Stream Bool) (x, y, c) fun ⟨x, y, c⟩ => Id.run <| do
+    match x 0, y 0, c 0 with
+      | none, _, some true => (none, (x.tail, y, c)) -- could not pop anything
+      | some _, _, some true => (x 0, (x.tail, y, c.tail)) -- pop from x
+      | _, none, some false => (none, (x, y.tail, c)) -- could not pop anything
+      | _, some _, some false => (y 0, (x, y.tail, c.tail)) -- pop from y
+      | _, _, none => (none, (x, y, c.tail)) -- no pop
+
+-- discards any data: actually this should not return anything
```

Comment:
Yeah in practice this should not return anything, but to be able to consume a stream you have to produce another stream.  But as long as the stream never blocks and can always consume a stream, then I think it has the behaviour that we are looking for.